node/network/bridge/src/multiplexer.rs

```diff
@@ -0,0 +1,135 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::convert::From;
+use std::pin::Pin;
+use std::result::Result;
+
+use futures::channel::mpsc;
+use futures::stream::Stream;
+use futures::task::{Context, Poll};
+use strum::IntoEnumIterator;
+
+use parity_scale_codec::{Decode, Error as DecodingError};
+
+use sc_network::config as network;
+use sc_network::PeerId;
+
+use polkadot_node_network_protocol::request_response::{
+	request::IncomingRequest, v1, Protocol, RequestResponseConfig,
+};
+use polkadot_subsystem::messages::AllMessages;
+
+/// Multiplex incoming network requests.
+///
+/// This multiplexer consumes all request streams and makes them a `Stream` of a single message
+/// type, useful for the network bridge to send them via the `Overseer` to other subsystems.
+pub struct RequestMultiplexer {
+	receivers: Vec<(Protocol, mpsc::Receiver<network::IncomingRequest>)>,
+	next_poll: usize,
+}
+
+/// Multiplexing can fail in case of invalid messages.
+pub struct RequestMultiplexError {
+	/// The peer that sent the invalid message.
+	pub peer: PeerId,
+	/// The error that occurred.
+	pub error: DecodingError,
```

Comment:
should timeout also be an `Error` variant?

Reply:
The multiplexer just handles incoming requests, it has no way of detecting the timeouts which happen in substrate. That a request is no longer valid can only be detected when sending back the response, where we would notice a no longer existing receiver.